File: algorithms/searching_algorithms/binary_search.py
# Binary Search in python


def binarySearch(array, x, low, high):

    # Repeat until the pointers low and high meet each other
    while low <= high:

        mid = low + (high - low)//2

        if array[mid] == x:
            return mid

        elif array[mid] < x:
            low = mid + 1

        else:
            high = mid - 1

    return -1



def binary_search_recursion(arr, value, low, hight):
    
    if low <= hight:
        
        mid = low + (hight - low)//2 
        
        if arr[mid]  == value:
            return mid 
        
        
        elif arr[mid] > value:
            return binary_search_recursion(arr, value, low, mid - 1)
        
        else: 
            return binary_search_recursion(arr, value, mid + 1, hight)
    else:
        
        return -1  


# Time Complexities

# Best case complexity: O(1)
# Average case complexity: O(log n)
# Worst case complexity: O(log n)
# Space Complexity

# The space complexity of the binary search is O(1).


array = [3, 4, 5, 6, 7, 8, 9]
x = 4



result = binary_search_recursion(array, x, 0, len(array)-1)

if result != -1:
    print("Element is present at index " + str(result))
else:
    print("Not found")
    
    
    
    
    

File: algorithms/C++/binary_search.cpp
#include <iostream>
#include <stdio.h>

using namespace std; 

int binary_search(int arr[], int value, int hight , int low){

    if (low <= hight){
        
        int mid = low + (hight - low)/2;
        if (arr[mid] == value){
            return 1;
        }
        else if (arr[mid] > value)
        {
            /* code */
            return binary_search(arr, value, mid - 1, low);
        }
        else{
            return binary_search(arr, value, hight, mid + 1);
        }
    
        
    }
    else{
        return -1;
    }
}



int main(void){
    int arr[] = {0, 2, 3, 4, 5, 6};
    int x = 4;
    int n = sizeof(arr) / sizeof(arr[0]);
    int result;
    result = binary_search(arr, x,n-1, 0);
    if (result == 1){
        cout<< "value is found!" << endl;
    }else{
        cout << "value is not found!" << endl;
    }
}